### PR TITLE
Room-specific settings

### PR DESCRIPTION
**Pull request by kszczek**
This PR adds room-specific settings, which are propagated to all room members (for now it's just the max distance before audio cuts off, but there are many more possibilities). The settings can be modified only by the room host, other members can only inspect them. The feature is mostly done, just needs some polishing. What do you think about this?

Server-side PR https://github.com/ottomated/CrewLink-server/pull/39

host host-settings client-settings

### CODE REVIEW

**Comment by ottomated** [comment]
Pog, I'll look into merging this

**Comment by Ornamus** [comment]
This looks great! Super happy to see this PR happen. Just want to clarify to make sure I understand how this works, because if this gets merged I'll immediately be working with it - `serverSettings` is the current copy of the server settings that is being used, and `userServerSettings` is a copy that the user (i.e. host) is changing the settings?

**Comment by kszczek** [comment]
> 
> 
> This looks great! Super happy to see this PR happen. Just want to clarify to make sure I understand how this works, because if this gets merged I'll immediately be working with it - `serverSettings` is the current server settings that is being used, and `userServerSettings` is a copy that the user (i.e. host) modifies when changing the settings - and then `userServerSettings` is then what's relayed to the other users?

Exactly, you've got it. I've renamed `serverSettings` to `lobbySettings` and `userServerSettings` to `localLobbySettings`, because I think those names make much more sense. I'll cover some more edge cases today (e.g. set localLobbySettings to lobbySettings for new host when host changes) and it will be ready unless you have some ideas.

**Comment by kszczek** [comment]
Ok, I'm done polishing this, it's a green light from me. I've changed the server's behavior, so that it stores just a dictionary of settings so that it doesn't have to be updated each time a setting is added to client. The clients are also backward-compatible which means that if the client doesn't recognize a setting, it will just ignore it instead of crashing and if it doesn't receive some of the settings, it will use the default value for them.

**Comment by ottomated** [comment]
@kszczek Opinions on transmitting this data over RTC versus over the websocket? I think that would be better if possible

**Comment by kszczek** [comment]
Haven't thought about that at first, but you're right that it's better, because if someone would want to, they could spoof being host either way, but this solution reduces load on server. I'll try implementing this.

**Comment by tystuyfzand** [comment]
That won't really protect people from spoofing that they're the host, it'd just reduce an event on the websocket side. I still think sending it over the websocket is a good idea, as the websocket pretty much handles coordination and sync of clients, while RTC is used for voice data.

Probably need some way (consensus system or something?) to confirm that a player is the host, depending on if the clients know who the lobby host is without being the host, but I assume that playerId == 0 would be right? It would let the server know they are who they say they are, at least.

**Comment by kszczek** [comment]
@tystuyfzand I think I have a solution for this. When client will be joining a room, he'll have to send his clientId, and the server will check if this clienId is unique for the room the client is trying to join. If it's not unique then client is trying to spoof someone and we drop the connection (if we drop the server connection he will not be able to estabilish any p2p connections).

**Comment by tystuyfzand** [comment]
Consider passing the playerId that set them with the settings back to each person too, then compare that to the hostId. If it doesn't match, then you can just drop the configuration and ignore it.

**Comment by OhMyGuus** [comment]
Just for the info the clientid that is used here is also located in the "PlayerControl" class as "OwnerId" at 0x1C  in crewlink you can get it at player->objectptr->0x1C.
This can help u identify the host as well if your not the host yourself.

File: src/main/hook-ti.ts [hunk 1]
```diff
@@ -14,7 +14,9 @@ export const IOffsets = t.iface([], {
 	'playerCount': t.array('number'),
 	'playerAddrPtr': 'number',
 	'exiledPlayerId': t.array('number'),
-	'gameCode': t.array('number'),
+  'gameCode': t.array('number'),
+  'hostId': t.array('number'),
+  'clientId': t.array('number'),
 	'player': t.iface([], {
```
Comment (DispatchCommit) [inline src/main/hook-ti.ts:20]: indentation is different here

Comment (kszczek) [reply src/main/hook-ti.ts:20]: Fixed in 4b3abc74fc22094cd7410a75cbfb8afbab9d6a8d

**Comment by ottomated** [comment]
Appreciate the PR. Next time please set it as a draft until it's done, there were several issues and just unimplemented things in here.

**Pull request merged**